book/02-basics-data.table.qmd

```diff
@@ -0,0 +1,39 @@
+# Quick data.table Intro for Beginners {#data.table}
+
+```{r 02-basics-R6-001, include = F}
+library(mlr3)
+library(mlr3book)
+```
+
+The package `r mlr3book::cran_pkg("data.table")` essentially implements the eponymous alternative to R's `data.frame()`, i.e. an object to store tabular data.
+
+:::{.callout-note}
+We decided to use `data.table()` because it is blazingly fast and scales quite well on bigger data.
+Many functions of `r mlr_pkg("mlr3")` return `data.table`s which can conveniently be subsetted or combined with other outputs.
+If you happen to not like the syntax or are feeling more comfortable with other tools, base `data.frame`s or `r cran_pkg("tibble")`/`r cran_pkg("dplyr")` is just a single `as.data.frame()` or `as_tibble()` away.
+:::
+
+Data tables can be constructed using the `data.table()` function (whose interface is similar to  to `data.frame()`) or by converting an object with `as.data.table()`.
+```{r}
+library("data.table")
+df = data.frame(x = 1:12, y = rep(letters[1:3], each = 4))
+dt = as.data.table(df)
+```
+Although both objects store the data identically in memory, they are considerably different in their operation.
+First, the index operator `[` works slightly different.
+For both objects, the first argument (`i`) is used to select rows and the second argument (`j`) is used to select columns.
+However, column names are in the search path, and thus can be used directly:
+```{r}
+df[df$y == "a", ]
+dt[y == "a", ]
+```
+
+Second, there is no optional `drop` argument (`drop` is always `FALSE` for `data.table()`), but instead multiple additional arguments to query the data from the `data.table()` in a very concise way.
+Most importantly, you can group the data with `by` and combine this with aggregating functions in `j`:
```

Comment:
Not clear what "in `j`" means here (there's no `j` in the code).

Reply:
yep, clarified this